### ISSUE TITLE
Response time is slow when there are a lot of handlers registered in MSW

### ISSUE THREAD

**Issue by Reverbe** [kettanaito/msw#1518]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 14 or higher

### Browsers

Chromium (Chrome, Brave, etc.)

### Reproduction repository

https://codesandbox.io/s/eloquent-sinoussi-7vuxkh?file=/src/mocks/handlers.js

### Reproduction steps

Please check the attached codesandbox sample where I've included the instructions.

To sum it up, it seems that having a high number of endpoints registered in MSW.js has a big impact on performance. 

Based on my tests:
<img width="813" alt="Pasted Graphic" src="https://user-images.githubusercontent.com/5213360/212078461-e0d03c54-0f16-4fcb-9299-e18d6f174b4a.png">

You have 2 variables to change in the codesandbox app to reproduce the issue.

Could it be that I have something misconfigured?

### Current behavior

We have a use case where we need to launch thousands of requests in parallel to the same endpoint, with different parameters. 

Since we use MSW not only for automated tests, but also for local manual testing / development, it becomes a killer of performance in the above scenario.

But not only this, as it can be see in the attached picture, even for individual requests, the performance hit is relatively big where we have a lot of handlers registered in MSW.js. We have about 1500 in our app.

### Expected behavior

I do expect to have a performance hit when having a large amount of endpoints registered, but not to this extent.

Again, perhaps I'm doing something wrong in configuring it. Please advise :)

**Comment by kettanaito** [comment]
Hey, @Reverbe. Thanks for raising this. I will elaborate on this behavior below.

## Why is MSW so slow when there's a large number of handlers?

When MSW captures a request, it performs what's referred to as "lookup"—the action of finding a relevant request handler for that request. The lookup is performed by iterating over _all the handlers_ left-to-right, finding that whose predicate matches the request, and then checking if that handler returns any mocked response. If it doesn't, skip it and move on to the next matching handler. 

Request handlers are stored in an Array internally, so the bigger is the array, the longer it takes to iterate over it. Besides, certain handlers may be asynchronous, which means MSW needs to await them fully before it can check whether they've handled this particular request in any way (e.g. in the case of conditional handlers that return nothing only for particular requests). 

## How to resolve this?

### Resource-based handlers

First of all, I'd look into the large number of handlers as it's likely unnecessary and is a result of inefficient MSW setup. Keep in mind that each request handler is meant to represent a _resource route on the server_. I doubt that your server has 1,500+ _distinct_ resource routes. At the same time, it's easy to fall into a trap of thinking about request handlers as _scenarios_:

```js
rest.get('/user/abc-123', resolver),
rest.get('/user/def-456', resolver),
rest.get('/user/xyz-890', resolver),
```

In the example above, there's no need to have 3 separate handlers for what is a `/user/:id` route. If you wish to reflect specific scenarios, like user ID-based responses, encapsulate them in a _single handler_ that describes the server resource:

```js
rest.get('/user/:id', (req, res, ctx) => {
  switch (req.params.id) {
    case 'abc-123': {
      return res(ctx.text('one'))
    }

    case 'def-456': {
      return res(ctx.text('two'))
    }

    default: {
      return res(ctx.text('fallback'))
    }
  }
})
```

Having your request handlers structured on a per-resource basis will improve MSW performance since it'd have less handlers to iterate through. 

### Runtime handlers

Another possible reason is that you're declaring request handlers that are not needed for the current execution context (read: unused handlers). This is solved by using runtime request handlers (`worker.use()`/`server.use()`) when appropriate instead of keeping the large number of handlers _all the time_ on the premise that they will be needed eventually. 

```js
// mocks/handlers.js
export const handlers = [...happyPathHandlers]
```

```js
// usage.js
import { handlers } from 'mocks/handlers'
import { setupServer } from 'msw/node'

setupServer(
  ...handlers,
  // Plus, any number of necessary runtime handlers here.
  rest.get('/user', resolver)
)
```

**Comment by kettanaito** [comment]
In your case, I'd recommend the resource-based handlers: try to find a pattern in your request handler URLs and generate, ideally, a small number of handlers that has all the repetitive logic _inside_ the handler. This should improve the performance of the lookup algorithm. 

We may look into improving the lookup algorithm but given it's a rather edge case issue (just as having 1,500 handlers), I won't put it on the roadmap yet.
